# Datagames 
Czechitas, 28.8.2021, Praha

# Kryptoměny

Svoje úspory máš ve třech kryptoměnách: Czechitacoinu, Polcoinu a PyCoinu. Protože ale počítač funguje jen částečně, nevidíš, kolik máš peněz, můžeš se ale podívat na všechny svoje transakce.

Proveď součet všech transakcí pro jednotlivé kryptoměny a následně převeď hodnotu všech úspor do kryptoměny Polcoin.

In [1]:
exchange_rates = {"PyCoin": 20.11, "Czechitacoin": 45.03}

In [2]:
import pandas as pd
df = pd.read_csv('transaction_list.csv', sep =';', header = None)
df.head()

0        1             2
0  2347-01-04  6609.00       Polcoin
1  2347-01-07   312.18        PyCoin
2  2347-01-07   341.37        PyCoin
3  2347-01-10     2.00  Czechitacoin
4  2347-01-11   -15.32        PyCoin

In [3]:
def polcoin(x):
    if x == "Polcoin":
        return 1
    elif x == "PyCoin":
        return exchange_rates["PyCoin"]
    elif x == "Czechitacoin":
        return exchange_rates["Czechitacoin"]
df[3]=df[2].apply(polcoin)
df.head()

0        1             2      3
0  2347-01-04  6609.00       Polcoin   1.00
1  2347-01-07   312.18        PyCoin  20.11
2  2347-01-07   341.37        PyCoin  20.11
3  2347-01-10     2.00  Czechitacoin  45.03
4  2347-01-11   -15.32        PyCoin  20.11

In [4]:
df["Polcoin"] = df[1] * df[3]
df["Polcoin"].sum()

230099.14840000003

# Hráči

Najdi jméno hráče na základě následujících informací.

Hráč byl aktivní v letech 2338 až 2345. Jeden z týmů, za které hrál, byl CHC.
Patří na seznam 100 nejlepších hráčů 30. let 24. století v počtu fanoušků na sociálních sítích.
V roce 2042 dal více než 5 gólů, měl méně asistencí než gólů a úspěšnost střelby měl více než 20 %.
Pilot kartičku koupil během svého posledního pobytu na stanici DX-359, což bylo od 30. 12. 2346 do 2. 1. 2347 a je si jistý, že kartičku zatím neprodal.

V archivu s daty najdeš následující soubory:

Soubor se seznamem kartiček, které pilot nakoupil. U každé kartičky eviduje datum nákupu, případný datum prodeje, cenu, za kterou kartičku koupil a případně prodal.
Soubor se statistikou hráčů ve 30. a 40. letech 24. století. Každý řádek obsahuje statistiky hráče v jednom roce. Statistiky obsahují ínformace o tom, za jaký tým hráč hrál, kolik odehrál zápasů, kolik dal gólů, kolik mě asistencí atd.
Seznam 100 hráčů s největším počtem fanoušků na sociálních sítích.

In [5]:
players = pd.read_json('players.json')
players.head()

BoughtOn            Player
0  2346-08-18       Megan Chang
1  2346-08-18      Robert Green
2  2346-06-02  William Sullivan
3  2346-08-18    Kristen Turner
4  2346-09-07      Thomas Silva

In [6]:
best_players = pd.read_csv('best_players.txt', header=None)
best_players.head()

0
0   Jorge Trujillo
1   Bryan Davidson
2      Shawn Giles
3       Kara Brown
4  Mallory Barnett

In [7]:
statistics = pd.read_json('statistics.json')
statistics.head()

year first_name last_name  shots  assistance  goals  matches team
0  2340      Megan     Chang     30           5      1       17  5MV
1  2341      Megan     Chang     29           6      2       12  0TT
2  2342      Megan     Chang      4           3      7       13  JBM
3  2343      Megan     Chang     27           6      6       19  CHC
4  2344      Megan     Chang      8           8      7       24  WLH

In [8]:
statistics['Player'] = statistics.first_name + ' ' + statistics.last_name
statistics.head()

year first_name last_name  shots  assistance  goals  matches team  \
0  2340      Megan     Chang     30           5      1       17  5MV   
1  2341      Megan     Chang     29           6      2       12  0TT   
2  2342      Megan     Chang      4           3      7       13  JBM   
3  2343      Megan     Chang     27           6      6       19  CHC   
4  2344      Megan     Chang      8           8      7       24  WLH   

        Player  
0  Megan Chang  
1  Megan Chang  
2  Megan Chang  
3  Megan Chang  
4  Megan Chang

In [9]:
df = pd.merge(statistics, players, on=['Player'], how='outer')
df.head()

year first_name last_name  shots  assistance  goals  matches team  \
0  2340      Megan     Chang     30           5      1       17  5MV   
1  2341      Megan     Chang     29           6      2       12  0TT   
2  2342      Megan     Chang      4           3      7       13  JBM   
3  2343      Megan     Chang     27           6      6       19  CHC   
4  2344      Megan     Chang      8           8      7       24  WLH   

        Player    BoughtOn  
0  Megan Chang  2346-08-18  
1  Megan Chang  2346-08-18  
2  Megan Chang  2346-08-18  
3  Megan Chang  2346-08-18  
4  Megan Chang  2346-08-18

In [10]:
df["success"] = (df.goals / df.shots)*100
df.head()

year first_name last_name  shots  assistance  goals  matches team  \
0  2340      Megan     Chang     30           5      1       17  5MV   
1  2341      Megan     Chang     29           6      2       12  0TT   
2  2342      Megan     Chang      4           3      7       13  JBM   
3  2343      Megan     Chang     27           6      6       19  CHC   
4  2344      Megan     Chang      8           8      7       24  WLH   

        Player    BoughtOn     success  
0  Megan Chang  2346-08-18    3.333333  
1  Megan Chang  2346-08-18    6.896552  
2  Megan Chang  2346-08-18  175.000000  
3  Megan Chang  2346-08-18   22.222222  
4  Megan Chang  2346-08-18   87.500000

In [11]:
df[(df.year > 2338) & (df.year < 2345) & (df.team == 'CHC') & ((df.year == 2342) & (df.goals > 5) & (df.assistance < df.goals) & (df.success > 20))]

year first_name last_name  shots  assistance  goals  matches team  \
71   2342     Rachel    Sutton      1           4      7       27  CHC   
129  2342    Mikayla   Sanchez     27           6      8       17  CHC   
171  2342      Linda      West     10           2      8       10  CHC   

              Player    BoughtOn    success  
71     Rachel Sutton  2346-05-18  700.00000  
129  Mikayla Sanchez  2347-01-02   29.62963  
171       Linda West  2346-01-06   80.00000

In [12]:
best_players[(best_players[0] == "Rachel Sutton")|(best_players[0] == "Linda West")|(best_players[0] == "Mikayla Sanchez")]

0
55  Mikayla Sanchez

# Němčina

Němčina rozlišuje tři rody - mužský, ženský a střední. Podstatná jména mužského rodu mají určitý člen "der", podstatná jména ženského rodu "die" a podstatná jména středního rodu "das". V množném čísle mají všechna přídavná jména určitý člen "die". Množné číslo se z jednotného čísla vytváří různými způsoby, např. některým podstatným jménům ženského rodu je na konec přidána koncovka "en" (nebo pouze "n", pokud tvar jednotného čísla končí na "e"). Pravidlo však není stoprocentní. 

Spočítej, kolik procent podstatných jmen ženského rodu se řídí popsaným pravidlem, tj. množné číslo vznikne přidáním koncovky "en" nebo "n". Nezapomeň do výpočtu zahrnout pouze podstatná jména v ženském rodě.

In [13]:
# wget.download('http://czechitas-data-games.herokuapp.com/media/media/validation_input.csv')

In [14]:
jmena = pd.read_csv('validation_input.csv',sep =';', header = None)
jmena.head()

0                   1
0  die Auslieferung  die Auslieferungen
1     die Botschaft     die Botschaften
2  die Vereinbarung  die Vereinbarungen
3      die Spannung      die Spannungen
4  die Gefolgschaft  die Gefolgschaften

In [15]:
jmena['clen']=jmena[0].str[:3]
jmena.head()

0                   1 clen
0  die Auslieferung  die Auslieferungen  die
1     die Botschaft     die Botschaften  die
2  die Vereinbarung  die Vereinbarungen  die
3      die Spannung      die Spannungen  die
4  die Gefolgschaft  die Gefolgschaften  die

In [16]:
filtered = jmena[jmena.clen == "die"]
filtered

0                   1 clen
0    die Auslieferung  die Auslieferungen  die
1       die Botschaft     die Botschaften  die
2    die Vereinbarung  die Vereinbarungen  die
3        die Spannung      die Spannungen  die
4    die Gefolgschaft  die Gefolgschaften  die
..                ...                 ...  ...
958          die Last          die Lasten  die
959        die Brücke         die Brücken  die
964    die Verkündung    die Verkündungen  die
979   die Überwachung   die Überwachungen  die
981         die Stufe          die Stufen  die

[458 rows x 3 columns]

In [17]:
en = filtered[filtered[1].str[-2:] == "en"]
en

0                   1 clen
0    die Auslieferung  die Auslieferungen  die
1       die Botschaft     die Botschaften  die
2    die Vereinbarung  die Vereinbarungen  die
3        die Spannung      die Spannungen  die
4    die Gefolgschaft  die Gefolgschaften  die
..                ...                 ...  ...
958          die Last          die Lasten  die
959        die Brücke         die Brücken  die
964    die Verkündung    die Verkündungen  die
979   die Überwachung   die Überwachungen  die
981         die Stufe          die Stufen  die

[434 rows x 3 columns]

In [18]:
434 * 100 / 458

94.75982532751091

# Nákup

V přiloženém souboru máš seznam, kde každý řádek představuje druhy zboží nakoupené jedním zákazníkem. Najdi dvojici potravin, která se v nich vyskytuje nejčastěji (například jestli nejčastěji kupují dohromady olej a maso, máso s pečivem atd.).

In [19]:
df = pd.read_csv('validation_input_grocery.csv', sep =';', header=None, names=range(21))
df.head()

0                                   1   \
0          Potatoes                Spirits and liqueurs   
1           Rice  Margarine and other vegetable fats   
2  Fresh whole milk                         Other meats   
3              Eggs                         Frozen fish   
4           Rice                          Lager beer   

                                   2                           3   \
0               Other bakery products      Confectionery products   
1                    Ready-made meals  Jams, marmalades and honey   
2                     Cheese and curd                Edible offal   
3                         Soft drinks   Edible ices and ice cream   
4  Margarine and other vegetable fats        Spirits and liqueurs   

                       4                             5   \
0             Frozen fish                     Chocolate   
1                   Sugar         Other bakery products   
2              Lager beer  Dried, salted or smoked meat   
3      Sauces, condiments                         Bread   
4  Confectionery products                         Pizza   

                                         6   \
0                           Cheese and curd   
1                      Dried fruit and nuts   
2  Dried, smoked or salted fish and seafood   
3                          Ready-made meals   
4                           Fortified wines   

                                   7                                   8   \
0  Margarine and other vegetable fats                             Poultry   
1                Other alcoholic beer                     Cheese and curd   
2                  Fresh low fat milk  Margarine and other vegetable fats   
3                        Edible offal                       Beef and veal   
4               Other bakery products                           Chocolate   

                    9   ...  \
0      Fortified wines  ...   
1    Other edible oils  ...   
2              Yoghurt  ...   
3      Fortified wines  ...   
4  Other milk products  ...   

                                                  11  \
0                                   Fresh whole milk   
1                                                NaN   
2                                   Wine from grapes   
3  Dried vegetables, other preserved or processed...   
4                                          Olive oil   

                                         12  \
0                    Fresh or chilled fruit   
1                                       NaN   
2  Preserved fruit and fruit-based products   
3                                       NaN   
4                      Other alcoholic beer   

                                         13                     14  \
0                            Preserved milk  Other cereal products   
1                                       NaN                    NaN   
2                                       NaN                    NaN   
3                                       NaN                    NaN   
4  Dried, smoked or salted fish and seafood       Wine from grapes   

                                         15      16           17          18  \
0  Dried, smoked or salted fish and seafood   Sugar         Eggs  Cigarettes   
1                                       NaN     NaN          NaN         NaN   
2                                       NaN     NaN          NaN         NaN   
3                                       NaN     NaN          NaN         NaN   
4                           Cheese and curd  Crisps  Frozen fish   Baby food   

                                                  19   20  
0                                                NaN  NaN  
1                                                NaN  NaN  
2                                                NaN  NaN  
3                                                NaN  NaN  
4  Other preserved or processed fish and seafood-...  NaN  

[5 rows x 21 columns]

In [20]:
import csv

with open('validation_input_grocery.csv') as f:
    reader = csv.reader(f, delimiter=';')
    df_list = list(reader)

df_list

[['Potatoes',
  'Spirits and liqueurs',
  'Other bakery products',
  'Confectionery products',
  'Frozen fish',
  'Chocolate',
  'Cheese and curd',
  'Margarine and other vegetable fats',
  'Poultry',
  'Fortified wines',
  'Yoghurt',
  'Fresh whole milk',
  'Fresh or chilled fruit',
  'Preserved milk',
  'Other cereal products',
  'Dried, smoked or salted fish and seafood',
  'Sugar',
  'Eggs',
  'Cigarettes'],
 ['Rice',
  'Margarine and other vegetable fats',
  'Ready-made meals',
  'Jams, marmalades and honey',
  'Sugar',
  'Other bakery products',
  'Dried fruit and nuts',
  'Other alcoholic beer',
  'Cheese and curd',
  'Other edible oils',
  'Preserved milk'],
 ['Fresh whole milk',
  'Other meats',
  'Cheese and curd',
  'Edible offal',
  'Lager beer',
  'Dried, salted or smoked meat',
  'Dried, smoked or salted fish and seafood',
  'Fresh low fat milk',
  'Margarine and other vegetable fats',
  'Yoghurt',
  'Spirits and liqueurs',
  'Wine from grapes',
  'Preserved fruit and 

In [21]:
from itertools import combinations
from collections import Counter

def most_freq(l, n):
    temp=[]
    for i in l:
        temp.extend(tuple(k) for k in combinations(i, n))
    res = Counter(temp)
    return res

In [22]:
tabulka = most_freq(df_list,2)
tabulka.most_common()[0]

(('Fresh or chilled vegetables other than potatoes and other tubers',
  'Olive oil'),
 5892)

# Karty

Dopočítej výsledky turnaje.

Karetní hra je trochu podobná pokeru. Každý hráč dostane 6 karet a každá karta je označena dvěma písmeny - barvou (H T, C, P - označují kříže, káry, srdce a píky) a hodnotou (čísla od 2 do 9 a písmeny J (Jack), Q (Queen), K (King) a A).

Hra má následující pravidla:

Každý hráč má 6 karet.
Vyhrává ten hráč, který má v kartách tu nejhodnotnější kombinace. Ten dostane 1 bod. Pokud mají dva hráč stejně hodnotnou kombinaci, získávají bod oba.
Kombinace seřazené od nejméně hodnotné po nejhodnotnější jsou:

    - dvě karty stejné hodnoty,
    - tři karty stejné hodnoty,
    - posloupnost pěti po sobě jdoucích karet (bez ohledu na barvu)
    - čtveřice karet stejné hodnoty.
Pokud nikdo z hráčů nezíská žádnou bodovanou kombinaci, nezískává bod nikdo.

U určení vítěze závisí pouze na tom, jakou kombinaci hráč má. Pokud například má jeden hráč dvě karty hodnoty 3 a druhý hráč má dvě karty hodnoty Q, získávají bod oba.

Turnaje se zůčastnili hráči Alice, Bob, Claire, Dee, a Elizabeth. Zjisti, který hráč měl na konci nejvíce bodů a kolik jich měl. Výsledek zapiš ve formátu Jméno Body, např. Claire 4.

V datovém souboru máš vždy karty, které měl na konci hry každý hráč v ruce. Jednotlivé karty jsou oddělené pomlčkou a hráči jsou odděleni středníkem, pořadí hráčů je dle abecedy.

Pořadí pro posloupnost karet je 23456789JQKA (tj. A se uznává pouze na úplném konci, nikoli na začátku).

In [23]:
df = pd.read_csv('validation_input_cards.csv', sep =';', index_col = False, names=['Alice', 'Bob', 'Claire', 'Dee', 'Elizabeth'])
df.head()

Alice                Bob             Claire                Dee  \
0  TJ-HJ-H7-P4-T3-P5  T7-C4-TA-TQ-C6-C9  H4-CK-PJ-CJ-CQ-T5  T6-T4-CA-P3-PQ-H3   
1  T2-CQ-C9-P6-C4-PK  H7-P9-H6-P5-C6-PA  T5-T9-T8-TK-HA-T3  H4-HQ-CJ-C2-H2-HK   
2  H5-TJ-C3-H8-HQ-H4  T4-P8-CK-CQ-T7-C8  P5-H7-PK-P4-C5-H2  PQ-T3-P2-H6-T5-P6   
3  TK-PK-CA-P3-HA-H8  T3-C4-T2-H6-H9-P6  PJ-P4-HK-T7-C7-PQ  H7-CJ-P8-CK-T5-C8   
4  P8-H5-P5-T3-TK-T8  C3-T2-C7-T5-H6-CA  C9-P7-CJ-C5-H3-HK  H4-C6-PJ-HA-HQ-T4   

           Elizabeth  
0  T2-P6-C7-P2-P7-H6  
1  C8-TA-H5-HJ-P8-P2  
2  PA-PJ-C7-P7-CJ-C2  
3  PA-TQ-H2-HJ-TJ-C3  
4  T7-H7-T6-TQ-PK-H2

In [24]:
def cards(vstup):
    mylist = vstup.split('-')
    karty=[]
    for i in mylist:
        karty.append(i[1])
    y=0
    for x in karty:
        if karty.count(x) > y:
            y=karty.count(x)
    if y == 2:
        return 1
    if y == 3:
        return 2
    if y == 4:
        return 4
    cisla = []
    for x in karty:
        if x == 'J':
            y = 10
        elif x == 'Q':
            y = 11
        elif x == 'K':
            y = 12
        elif x == 'A':
            y = 13
        else:
            y = int(x)
        cisla.append(y)
    
    cisla = sorted(cisla)
    z=0
    for x in range(1,6):
        if cisla[x] == cisla[x-1] + 1:
            z += 1  
    if z >= 4:
        return 3
    else:
        return 0

In [25]:
df["Alice"] = df["Alice"].apply(cards)
df["Bob"] = df["Bob"].apply(cards)
df["Claire"] = df["Claire"].apply(cards)
df["Dee"] = df["Dee"].apply(cards)
df["Elizabeth"] = df["Elizabeth"].apply(cards)
df

Alice  Bob  Claire  Dee  Elizabeth
0       1    0       1    1          1
1       0    1       0    1          1
2       0    1       1    1          1
3       1    1       1    1          1
4       1    0       0    1          1
5       2    2       1    1          1
6       1    1       1    1          2
7       0    0       2    0          1
8       1    1       1    1          0
9       1    1       0    3          0
10      1    0       1    0          1
11      1    1       1    1          2
12      0    0       1    1          1
13      1    1       2    1          1
14      1    1       1    1          1
15      0    0       0    0          0
16      0    1       1    1          1
17      1    1       1    1          3
18      1    0       1    0          1
19      0    1       0    1          1

In [26]:
Alice = df.Alice.sum()
print(f"Alice: {Alice}")
Bob = df.Bob.sum()
print(f"Bob: {Bob}")
Claire = df.Claire.sum()
print(f"Claire: {Claire}")
Dee = df.Dee.sum()
print(f"Dee: {Dee}")
Elizabeth = df.Elizabeth.sum()
print(f"Elizabeth: {Elizabeth}")

Alice: 14
Bob: 14
Claire: 17
Dee: 18
Elizabeth: 21
